src/proxyp/OutgoingHttpConfig.cc

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "acl/FilledChecklist.h"
+#include "acl/Gadgets.h"
+#include "acl/Tree.h"
+#include "base/TextException.h"
+#include "cache_cf.h"
+#include "ConfigOption.h"
+#include "ConfigParser.h"
+#include "parser/Tokenizer.h"
+#include "proxyp/Header.h"
+#include "proxyp/OutgoingHttpConfig.h"
+#include "sbuf/Stream.h"
+#include "sbuf/StringConvert.h"
+
+ProxyProtocol::Option::Option(const char *aName, const char *aVal, bool quoted)
+    : theName(aName), theValue(aVal), valueFormat(nullptr)
+{
+    if (quoted)
+        parseFormat();
+}
+
+void
+ProxyProtocol::Option::dump(std::ostream &os)
+{
+    os << theName << '=' << ConfigParser::QuoteString(SBufToString(theValue));
+}
+
+void
+ProxyProtocol::Option::parseFormat()
+{
+    Assure(!valueFormat);
+    valueFormat = new Format::Format(theName.c_str());
+    if (!valueFormat->parse(theValue.c_str())) {
+        delete valueFormat;
+        throw TextException(ToSBuf("failed to parse value ", theValue), Here());
+    }
+}
+
+SBuf
+ProxyProtocol::Option::processFormat(const AccessLogEntryPointer &al) const
+{
+    Assure(valueFormat);
+    static MemBuf mb;
+    mb.reset();
+    valueFormat->assemble(mb, al, 0);
+    return SBuf(mb.content());
+}
+
+static std::optional<Ip::Address>
+ParseAddr(const SBuf &val)
+{
+    const auto addr = Ip::Address::Parse(SBuf(val).c_str());
+    if (!addr)
+        throw TextException(ToSBuf("Cannot parse '", val, "' as an IP address"), Here());
+    return addr;
+}
+
+ProxyProtocol::AddrOption::AddrOption(const char *aName, const char *aVal, bool quoted) : Option(aName, aVal, quoted)
+{
+    if (!valueFormat || !valueFormat->hasPercentCode())
```

Comment:
Please see if we can refactor so that we do not store valueFormat when it is used only once (here). Same for storing theValue when valueFormat has %codes, actually. In other words, try to avoid storing anything we are not going to use after construction.